File: src/extractors/haskell.ts
import type {
  ExtractorOutput,
  SubDeclaration,
  TreeSitterNode,
  TreeSitterTree,
} from '../types.js';
import { findChild, nodeEndLine } from './helpers.js';

/**
 * Extract symbols from Haskell files.
 *
 * Note: tree-sitter-haskell uses `type_synomym` (misspelled) for type aliases.
 */
export function extractHaskellSymbols(tree: TreeSitterTree, _filePath: string): ExtractorOutput {
  const ctx: ExtractorOutput = {
    definitions: [],
    calls: [],
    imports: [],
    classes: [],
    exports: [],
    typeMap: new Map(),
  };

  walkHaskellNode(tree.rootNode, ctx);
  return ctx;
}

function walkHaskellNode(node: TreeSitterNode, ctx: ExtractorOutput): void {
  switch (node.type) {
    case 'function':
      handleHaskellFunction(node, ctx);
      break;
    case 'bind':
      handleHaskellBind(node, ctx);
      break;
    case 'data_type':
      handleHaskellDataType(node, ctx);
      break;
    case 'newtype':
      handleHaskellNewtype(node, ctx);
      break;
    case 'type_synomym':
      handleHaskellTypeSynonym(node, ctx);
      break;
    case 'class':
      handleHaskellClass(node, ctx);
      break;
    case 'instance':
      handleHaskellInstance(node, ctx);
      break;
    case 'import':
      handleHaskellImport(node, ctx);
      break;
    case 'apply':
      handleHaskellApply(node, ctx);
      break;
  }

  for (let i = 0; i < node.childCount; i++) {
    const child = node.child(i);
    if (child) walkHaskellNode(child, ctx);
  }
}

function handleHaskellFunction(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  const params = extractHaskellParams(node);

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'function',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: params.length > 0 ? params : undefined,
  });
}

function extractHaskellParams(funcNode: TreeSitterNode): SubDeclaration[] {
  const params: SubDeclaration[] = [];
  // Haskell function patterns are positional children
  for (let i = 0; i < funcNode.childCount; i++) {
    const child = funcNode.child(i);
    if (!child) continue;
    if (child.type === 'patterns' || child.type === 'parameter') {
      for (let j = 0; j < child.childCount; j++) {
        const pat = child.child(j);
        if (pat && (pat.type === 'variable' || pat.type === 'identifier')) {
          params.push({ name: pat.text, kind: 'parameter', line: pat.startPosition.row + 1 });
        }
      }
    }
    if (child.type === 'variable' && i > 0) {
      // Pattern parameters after the function name
      params.push({ name: child.text, kind: 'parameter', line: child.startPosition.row + 1 });
    }
  }
  return params;
}

function handleHaskellBind(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'variable',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function handleHaskellDataType(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;
  const name = nameNode.text;

  const children: SubDeclaration[] = [];
  // Extract constructors
  const constructors = node.childForFieldName('constructors');
  if (constructors) {
    for (let i = 0; i < constructors.childCount; i++) {
      const ctor = constructors.child(i);
      if (!ctor) continue;
      if (ctor.type === 'data_constructor' || ctor.type === 'gadt_constructor') {
        const ctorName = findChild(ctor, 'constructor') || findChild(ctor, 'constructor_operator');
        if (ctorName) {
          children.push({
            name: ctorName.text,
            kind: 'property',
            line: ctor.startPosition.row + 1,
          });
        }
      }
    }
  }

  ctx.definitions.push({
    name,
    kind: 'type',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
    children: children.length > 0 ? children : undefined,
  });
}

function handleHaskellNewtype(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'type',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function handleHaskellTypeSynonym(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'type',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function handleHaskellClass(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'class',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function handleHaskellInstance(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const nameNode = node.childForFieldName('name');
  if (!nameNode) return;

  ctx.definitions.push({
    name: nameNode.text,
    kind: 'class',
    line: node.startPosition.row + 1,
    endLine: nodeEndLine(node),
  });
}

function handleHaskellImport(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const moduleNode = node.childForFieldName('module');
  if (!moduleNode) return;

  const source = moduleNode.text;
  const names: string[] = [];

  const alias = node.childForFieldName('alias');
  if (alias) names.push(alias.text);

  const importList = node.childForFieldName('names');
  if (importList) {
    for (let i = 0; i < importList.childCount; i++) {
      const item = importList.child(i);
      if (
        item &&
        (item.type === 'variable' || item.type === 'constructor' || item.type === 'type')
      ) {
        names.push(item.text);
      }
    }
  }

  ctx.imports.push({
    source,
    names: names.length > 0 ? names : [source.split('.').pop() || source],
    line: node.startPosition.row + 1,
  });
}

function handleHaskellApply(node: TreeSitterNode, ctx: ExtractorOutput): void {
  const funcNode = node.childForFieldName('function');
  if (!funcNode) return;

  // Only record named function applications, not complex expressions
  if (
    funcNode.type === 'variable' ||
    funcNode.type === 'constructor' ||
    funcNode.type === 'identifier'
  ) {
    ctx.calls.push({ name: funcNode.text, line: node.startPosition.row + 1 });
  } else if (funcNode.type === 'qualified_variable' || funcNode.type === 'qualified_constructor') {
    ctx.calls.push({ name: funcNode.text, line: node.startPosition.row + 1 });
  }
}

File: crates/codegraph-core/src/extractors/haskell.rs
use tree_sitter::{Node, Tree};
use crate::cfg::build_function_cfg;
use crate::complexity::compute_all_metrics;
use crate::types::*;
use super::helpers::*;
use super::SymbolExtractor;

pub struct HaskellExtractor;

impl SymbolExtractor for HaskellExtractor {
    fn extract(&self, tree: &Tree, source: &[u8], file_path: &str) -> FileSymbols {
        let mut symbols = FileSymbols::new(file_path.to_string());
        walk_tree(&tree.root_node(), source, &mut symbols, match_haskell_node);
        walk_ast_nodes_with_config(&tree.root_node(), source, &mut symbols.ast_nodes, &HASKELL_AST_CONFIG);
        symbols
    }
}

fn match_haskell_node(node: &Node, source: &[u8], symbols: &mut FileSymbols, _depth: usize) {
    match node.kind() {
        "function" => handle_haskell_function(node, source, symbols),
        "bind" => handle_haskell_bind(node, source, symbols),
        "data_type" => handle_haskell_data_type(node, source, symbols),
        "newtype" => handle_haskell_newtype(node, source, symbols),
        "type_synomym" => handle_haskell_type_synonym(node, source, symbols),
        "class" => handle_haskell_class(node, source, symbols),
        "instance" => handle_haskell_instance(node, source, symbols),
        "import" => handle_haskell_import(node, source, symbols),
        "apply" => handle_haskell_apply(node, source, symbols),
        _ => {}
    }
}

fn handle_haskell_function(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "function".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: compute_all_metrics(node, source, "haskell"),
        cfg: build_function_cfg(node, "haskell", source),
        children: None,
    });
}

fn handle_haskell_bind(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "variable".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_haskell_data_type(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };
    let name = node_text(&name_node, source).to_string();

    let mut children = Vec::new();
    if let Some(constructors) = node.child_by_field_name("constructors") {
        for i in 0..constructors.child_count() {
            if let Some(ctor) = constructors.child(i) {
                if ctor.kind() == "data_constructor" || ctor.kind() == "gadt_constructor" {
                    let ctor_name = find_child(&ctor, "constructor")
                        .or_else(|| find_child(&ctor, "constructor_operator"));
                    if let Some(cn) = ctor_name {
                        children.push(child_def(
                            node_text(&cn, source).to_string(),
                            "property",
                            start_line(&ctor),
                        ));
                    }
                }
            }
        }
    }

    symbols.definitions.push(Definition {
        name,
        kind: "type".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: opt_children(children),
    });
}

fn handle_haskell_newtype(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "type".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_haskell_type_synonym(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "type".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_haskell_class(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "class".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_haskell_instance(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let name_node = match node.child_by_field_name("name") {
        Some(n) => n,
        None => return,
    };

    symbols.definitions.push(Definition {
        name: node_text(&name_node, source).to_string(),
        kind: "class".to_string(),
        line: start_line(node),
        end_line: Some(end_line(node)),
        decorators: None,
        complexity: None,
        cfg: None,
        children: None,
    });
}

fn handle_haskell_import(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let module_node = match node.child_by_field_name("module") {
        Some(n) => n,
        None => return,
    };

    let source_name = node_text(&module_node, source).to_string();
    let mut names = Vec::new();

    if let Some(alias) = node.child_by_field_name("alias") {
        names.push(node_text(&alias, source).to_string());
    }

    if let Some(import_list) = node.child_by_field_name("names") {
        for i in 0..import_list.child_count() {
            if let Some(item) = import_list.child(i) {
                match item.kind() {
                    "variable" | "constructor" | "type" => {
                        names.push(node_text(&item, source).to_string());
                    }
                    _ => {}
                }
            }
        }
    }

    if names.is_empty() {
        let last = source_name.split('.').last().unwrap_or(&source_name).to_string();
        names.push(last);
    }

    symbols.imports.push(Import::new(source_name, names, start_line(node)));
}

fn handle_haskell_apply(node: &Node, source: &[u8], symbols: &mut FileSymbols) {
    let func_node = match node.child_by_field_name("function") {
        Some(n) => n,
        None => return,
    };

    match func_node.kind() {
        "variable" | "constructor" | "identifier" | "qualified_variable" | "qualified_constructor" => {
            symbols.calls.push(Call {
                name: node_text(&func_node, source).to_string(),
                line: start_line(node),
                dynamic: None,
                receiver: None,
            });
        }
        _ => {}
    }
}
